### PR TITLE
Improve performance for SYCL parallel_reduce

### PR DESCRIPTION
**Pull request by masterleinad**
- let each thread load two values
- use better memory access pattern
~~- let the last block do the final reduction~~

### CODE REVIEW

**Comment by masterleinad** [comment]
Benchmark showed better performance for the recursive implementation so I dropped the commit related to the one that lets the last block do the final reduction. A backup for that implementation can be found at https://github.com/masterleinad/kokkos/tree/sycl_reduce_performance_backup.

File: core/src/SYCL/Kokkos_SYCL_Parallel_Reduce.hpp [hunk 1]
```diff
@@ -217,64 +216,70 @@ class ParallelReduce<FunctorType, Kokkos::RangePolicy<Traits...>, ReducerType,
...
+              const auto local_id = item.get_local_linear_id();
+              const auto global_id =
+                  wgroup_size * item.get_group_linear_id() * values_per_thread +
+                  local_id;
 
               // Initialize local memory
+              const typename Policy::index_type upper_bound =
+                  (global_id + values_per_thread * wgroup_size) < size
+                      ? global_id + values_per_thread * wgroup_size
+                      : size;
```
Comment (dalg24) [inline core/src/SYCL/Kokkos_SYCL_Parallel_Reduce.hpp:228]: `std::min` might be more readable than this ternary op

**Review by dalg24 — APPROVED** [review]
It is hard to parse what is going on in this reduction algorithm :/

File: core/src/SYCL/Kokkos_SYCL_Parallel_Reduce.hpp (hunk 1 above)
Comment (masterleinad) [reply core/src/SYCL/Kokkos_SYCL_Parallel_Reduce.hpp:228]: Yes, I wasn't quite sure if that's supported on the device. I can check.

File: core/src/SYCL/Kokkos_SYCL_Parallel_Reduce.hpp [hunk 2]
```diff
@@ -168,23 +168,17 @@ class ParallelReduce<FunctorType, Kokkos::RangePolicy<Traits...>, ReducerType,
...
+    std::size_t size         = policy.end() - policy.begin();
+    const auto init_size     = std::max<std::size_t>(
+        ((size + 1) / 2 + wgroup_size - 1) / wgroup_size, 1);
     const auto value_count =
         FunctorValueTraits<ReducerTypeFwd, WorkTagFwd>::value_count(
             selected_reducer);
     const auto results_ptr =
         static_cast<pointer_type>(Experimental::SYCLSharedUSMSpace().allocate(
             "SYCL parallel_reduce result storage",
             sizeof(*m_result_ptr) * std::max(value_count, 1u) * init_size));
```
[the remark this answers is not in the document]
Comment (masterleinad) [reply core/src/SYCL/Kokkos_SYCL_Parallel_Reduce.hpp:184]: > It is hard to parse what is going on in this reduction algorithm :/

I thought I didn't change it too much but I'm happy to add some more comments.

**Comment by masterleinad** [comment]
> It is hard to parse what is going on in this reduction algorithm :/

I thought I didn't change it too much but I'm happy to add some more comments.

File: core/src/SYCL/Kokkos_SYCL_Parallel_Reduce.hpp [hunk 3]
```diff
@@ -168,23 +168,20 @@ class ParallelReduce<FunctorType, Kokkos::RangePolicy<Traits...>, ReducerType,
...
+        ((size + values_per_thread - 1) / values_per_thread + wgroup_size - 1) /
+            wgroup_size,
+        1);
     const auto value_count =
         FunctorValueTraits<ReducerTypeFwd, WorkTagFwd>::value_count(
             selected_reducer);
     const auto results_ptr =
         static_cast<pointer_type>(Experimental::SYCLSharedUSMSpace().allocate(
             "SYCL parallel_reduce result storage",
             sizeof(*m_result_ptr) * std::max(value_count, 1u) * init_size));
```
Comment (masterleinad) [inline core/src/SYCL/Kokkos_SYCL_Parallel_Reduce.hpp:184]: Yes, `sizeof(value_type)` should be the same. Let me try that.

File: core/src/SYCL/Kokkos_SYCL_Parallel_Reduce.hpp [hunk 4]
```diff
@@ -168,25 +168,24 @@ class ParallelReduce<FunctorType, Kokkos::RangePolicy<Traits...>, ReducerType,
     sycl::queue& q = *instance.m_queue;
 
     // FIXME_SYCL optimize
-    constexpr size_t wgroup_size = 32;
-    std::size_t size             = policy.end() - policy.begin();
-    const auto init_size =
-        std::max<std::size_t>((size + wgroup_size - 1) / wgroup_size, 1);
+    constexpr size_t wgroup_size       = 128;
+    constexpr size_t values_per_thread = 2;
```
Comment (crtrott) [inline core/src/SYCL/Kokkos_SYCL_Parallel_Reduce.hpp:172]: this values per thread, is probably right for something like dot product, but probably wrong for LAMMPS where the reduction is on rather beefy kernels which often don't have enough parallelism (think 30k atoms, each looping serially over its neighbors and then do a reduction over all of them) not sure if you care right now

**Review by crtrott — APPROVED** [review]
I think the values per thread thing is good for dot but bad for physics codes with reducitons
